### ISSUE TITLE
v0.14.2 crashes a few minutes after start

### ISSUE THREAD

**Issue by kpuc** [syncthing/syncthing#3448]
This is only happening on one Windows 7 x64 node at work, so far that I'm aware.  I've restarted syncthing twice and each panic appears very similar. 

[JABL6] 15:27:51 INFO: syncthing v0.14.2 "Dysprosium Dragonfly" (go1.6.3 windows-amd64) WIN-P6RP886TVDJ\Administrator@WIN-P6RP886TVDJ 2016-07-26 11:55:25 UTC
[JABL6] 15:27:51 INFO: My ID: JABL6VC

...  several minutes pass ... 
```
[JABL6] 15:39:04 INFO: Connection to RTBSNQZ closed: switching connections
[JABL6] 15:39:04 INFO: Established secure connection to RTBSNQZ at 10.0.1.26:49394-73.202.85.101:44109 (TCP (Client))
[JABL6] 15:39:04 INFO: Device RTBSNQZ client is "syncthing v0.14.2" named "lela-current.dax"
[JABL6] 15:39:05 INFO: Connection to RTBSNQZ closed: reading message: read tcp 10.0.1.26:49382->173.230.149.23:22067: use of closed network connection
[JABL6] 15:39:05 INFO: Connection to RTBSNQZ closed: reading length: read tcp 10.0.1.26:49394->73.202.85.101:44109: use of closed network connection
[JABL6] 15:39:05 INFO: Device RTBSNQZ folder "tunes_mp3" is delta index compatible (mlv=0)
[JABL6] 15:39:05 INFO: Device RTBSNQZ folder "SharedTools" is delta index compatible (mlv=0)
panic: runtime error: invalid memory address or nil pointer dereference
[JABL6] 15:39:05 INFO: Device RTBSNQZ folder "Eric2ChrisTest" is delta index compatible (mlv=2815)
[signal 0xc0000005 code=0x0 addr=0x38 pc=0x791490]
goroutine 3269 [running]:
panic(0xbf65e0, 0xc082008070)
	c:/go/src/runtime/panic.go:481 +0x3f4
github.com/syncthing/syncthing/lib/connections.(*Connection).ID(0xc08983c030, 0x0, 0x0, 0x0, 0x0)
	<autogenerated>:95 +0x60
github.com/syncthing/syncthing/lib/model.sendIndexes(0x37b3758, 0xc08983c030, 0xc08355aad0, 0xe, 0xc089ce2500, 0xc082916460, 0xaff)
	c:/jenkins/workspace/syncthing-release-windows/src/github.com/syncthing/syncthing/lib/model/model.go:1207 +0x43
created by github.com/syncthing/syncthing/lib/model.(*Model).ClusterConfig
	c:/jenkins/workspace/syncthing-release-windows/src/github.com/syncthing/syncthing/lib/model/model.go:733 +0x207a
```

**Comment by AudriusButkevicius** [comment]
I think the signal is a moot point.

The other peer disconnects before the sendIndexes loop starts and the ID check blows up?

**Comment by kpuc** [comment]
Panic-ed three times just a few minutes after starting.  Started Syncthing again after I posted this issue; this time it has been fine for just under two hours.   Not sure how to reproduce.

**Comment by calmh** [comment]
#3432 this feels like it may be a Go 1.6.3 thing since it started recently, I haven't filed it onward yet

**Comment by calmh** [comment]
@kpuc can you reproduce this with `STTRACE=connections,model`?

**Comment by AudriusButkevicius** [comment]
@kpuc was one of the devices paused remotely?

**Comment by paulhargreaves** [comment]
Potentially identical issue: https://forum.syncthing.net/t/panic-runtime-error-invalid-memory-address-or-nil-pointer-dereference/7892?u=uniquenospacesshort

None of my devices are paused.

**Comment by kpuc** [comment]
Added STTRACE=connections,model and we'll see how long it takes to crash.. 
Also crashed with very similar looking panic message on my FreeBSD node last night.
I don't think any nodes either of my machines that have seen this panic are paused, but I don't control all the nodes either machine connects to.  None of them were paused on _my_ side.

**Comment by calmh** [comment]
@kpuc did you have the issue on any of your other devices that they crashed due to missing or lacking space in /tmp ?

**Comment by kpuc** [comment]
Not that i'm aware of, no.

**Comment by calmh** [comment]
Never mind, I think the cause in your case is the switching connections thing. The fix I've written up above should work for you as well.

**Comment by kpuc** [comment]
It's happening far less often with v0.14.3 (Thanks!), but still occurs: this node is win10pro amd64.
Looks like it was switching between relay and direct, or the other way around.  or both. 

[JQHVO] 10:06:22 INFO: Connection to FYOHSB2 closed: switching connections
[JQHVO] 10:06:22 INFO: Established secure connection to FYOHSB2 at 192.168.181.10:22000-<FYOHSB2's public IP over the internet>:58180 (TCP (Server))
[JQHVO] 10:06:22 INFO: Connection to FYOHSB2 closed: reading length: read tcp 192.168.181.10:50169->104.129.57.94:22067: use of closed network connection
[JQHVO] 10:06:22 INFO: Device FYOHSB2 client is "syncthing v0.14.3" named <FYOHSB2>
[JQHVO] 10:06:22 INFO: Connection to FYOHSB2 closed: reading length: read tcp 192.168.181.10:22000->68.65.72.186:58180: use of closed network connection
[JQHVO] 10:06:45 INFO: Established secure connection to FYOHSB2 at 192.168.181.10:50186-104.129.57.94:22067 (Relay (Server))
[JQHVO] 10:06:45 INFO: Device FYOHSB2 client is "syncthing v0.14.3" named <FYOHSB2>
[JQHVO] 10:06:45 INFO: Device FYOHSB2 folder "9ZTJu-VPHqJ" is delta index compatible (mlv=75283)
[JQHVO] 10:06:45 INFO: Device FYOHSB2 folder "SharedTools" is delta index compatible (mlv=69294)
[JQHVO] 10:06:45 INFO: Device FYOHSB2 folder "randomBulkSync" is delta index compatible (mlv=69414)
[JQHVO] 10:06:45 INFO: Device FYOHSB2 folder "randomMinSync" is delta index compatible (mlv=72906)
[JQHVO] 10:06:45 INFO: Device FYOHSB2 folder "cfzjw-SkHX7" is delta index compatible (mlv=70180)
[JQHVO] 10:06:45 INFO: Device FYOHSB2 folder "DABingeNeedRetire" is delta index compatible (mlv=76601)
[JQHVO] 10:07:10 INFO: Connection to FYOHSB2 closed: switching connections
panic: bug: ClusterConfig called on closed or nonexistent connection
goroutine 8184 [running]:
[JQHVO] 10:07:10 INFO: Established secure connection to FYOHSB2 at 192.168.181.10:22000-<FYOHSB2's public IP over the internet>:24653 (TCP (Server))
panic(0xa9b5c0, 0xc08543e870)
[JQHVO] 10:07:10 INFO: Device FYOHSB2 client is "syncthing v0.14.3" named <FYOHSB2>
	c:/go/src/runtime/panic.go:481 +0x3f4
[JQHVO] 10:07:10 INFO: Connection to FYOHSB2 closed: writing message: write tcp 192.168.181.10:50186->104.129.57.94:22067: use of closed network connection
github.com/syncthing/syncthing/lib/model.(*Model).ClusterConfig(0xc082050140, 0xeb96024f07791c2e, 0xf66cc66753f4d053, 0x9aa6d3756b1bc563, 0xcd823ef98efc703b, 0xc082366800, 0x6, 0x8)
	c:/jenkins/workspace/syncthing-release-windows/src/github.com/syncthing/syncthing/lib/model/model.go:655 +0x214
github.com/syncthing/syncthing/lib/protocol.(*nativeModel).ClusterConfig(0xc0869c5620, 0xeb96024f07791c2e, 0xf66cc66753f4d053, 0x9aa6d3756b1bc563, 0xcd823ef98efc703b, 0xc082366800, 0x6, 0x8)
	<autogenerated>:59 +0x8c
github.com/syncthing/syncthing/lib/protocol.(*rawConnection).readerLoop(0xc0825f2c00, 0x0, 0x0)
	c:/jenkins/workspace/syncthing-release-windows/src/github.com/syncthing/syncthing/lib/protocol/protocol.go:305 +0x12d8
created by github.com/syncthing/syncthing/lib/protocol.(*rawConnection).Start
	c:/jenkins/workspace/syncthing-release-windows/src/github.com/syncthing/syncthing/lib/protocol/protocol.go:170 +0x40
